### PR TITLE
Feature/clear command

### PR DESCRIPTION
**Pull request by Juminstock**
Added clear command and documentation.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/backlog-ticket.md
Comment (codespool) [inline .github/ISSUE_TEMPLATE/backlog-ticket.md:1]: No need to edit the template, it's meant to be copied to the PR description.

File: src/commands/clear/index.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    } catch (err) {
+      throw new FileError(
+        `Error while to delete the folder ${path}: ${err}Error while to delete the folder ${path}: ${err}`
+      );
+    }
+  } 
+
+  public async run(): Promise<any> {
+
+    const spinner = new Spinner();
```
Comment (codespool) [inline src/commands/clear/index.ts:40]: there's a `spinner` instance in the parent class already.

File: src/commands/clear/index.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      description: "Name of the contract artifact to clear",
+    }),
+  };
+
+  static async deleteFolder(path: string): Promise<void> {
+    try {
+      await fs.remove(path);
+    } catch (err) {
+      throw new FileError(
+        `Error while to delete the folder ${path}: ${err}Error while to delete the folder ${path}: ${err}`
```
Comment (codespool) [inline src/commands/clear/index.ts:33]: Why the repetition?
When rethrowing, the orginal error can be included as a second arguement's `cause` property.
`throw new FileError("Error deleting the folder ${path}.", {cause: err})`

File: src/commands/clear/index.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+  static args = {
+    contractName: Args.string({
+      name: "contractName",
+      required: false,
+      description: "Name of the contract artifact to clear",
+    }),
+  };
+
+  static async deleteFolder(path: string): Promise<void> {
```
Comment (codespool) [inline src/commands/clear/index.ts:28]: Does this need to be static? What's the benefit? If it's meant to be used by other classes, it should be a part of `lib`

File: src/commands/clear/index.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import { Spinner } from "../../lib/index.js";
+import { SwankyCommand } from "../../lib/swankyCommand.js";
+import { ConfigError, FileError } from "../../lib/errors.js";
+import fs from "fs-extra"
+import path from "path"
+import { Args, Flags } from "@oclif/core";
+import { flagUsage } from "@oclif/core/lib/parser/help.js";
```
Comment (codespool) [inline src/commands/clear/index.ts:7]: Where is this used?

File: src/commands/clear/index.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    const spinner = new Spinner();
+
+    const { flags, args } = await this.parse(Clear);
+
+    if(!flags.all && !args.contractName) throw new ConfigError("You need to send any flag or argument.")
+
+    if(flags.all && args.contractName) {
+      throw new ConfigError("You can't sent the flag '-a' or '--all' and argument in the same time.")
+    }
```
Comment (codespool) [inline src/commands/clear/index.ts:48]: Checking only for `args.contractName === undefined && !flags.all` would cover the use case (if --all is passed, just ignore the arg)

File: src/commands/clear/index.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      }, {
+        name: "Target",
+        path: path.join(workDirectory, './target')
+      }, {
+        name: "Test",
+        contractName: args.contractName,
+        path: path.join(workDirectory, './tests/', args.contractName, "/artifacts")
+      })
+  
+      foldersToDelete.map(async (folder) => {
```
Comment (codespool) [inline src/commands/clear/index.ts:87]: Don't use `.map()` if you're not expecting a return. (use forEach() in that case)

File: src/commands/clear/index.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  
+      if (fs.existsSync(folder.path)) {
```
Comment (codespool) [inline src/commands/clear/index.ts:89]: `fs-extra`'s `.remove()` method doesn't need this check.

File: src/commands/clear/index.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  
+        this.log(`Deleting the ${folder.name} folder`)
+  
+        // await spinner.runCommand(async () => 
+        await Clear.deleteFolder(folder.path)
```
Comment (codespool) [inline src/commands/clear/index.ts:94]: the `this.spiner` instance has a `.runCommand()` for this use

**Comment by Juminstock** [comment]
Thanks for the feedback, Codespool. The PR has been updated based on your comments.

File: src/commands/clear/index.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        name: "Target",
+        path: path.join(workDirectory, './target')
+      }, {
+        name: "TestArtifacts",
+        contractName: args.contractName,
+        path: path.join(workDirectory, './tests/', args.contractName, "/artifacts")
+      })
+
+      for await (const folder of foldersToDelete) {
+        let resultSpinner = await this.spinner.runCommand( async () => this.deleteFolder(folder.path),
```
Comment (ipapandinas) [inline src/commands/clear/index.ts:76]: The variable `resultSpinner` is useless, no need to store the command result.

File: src/commands/clear/index.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+      }, {
+        name: "Target",
+        path: path.join(workDirectory, './target')
+      }, {
+        name: "TestArtifacts",
+        contractName: args.contractName,
+        path: path.join(workDirectory, './tests/', args.contractName, "/artifacts")
+      })
+
+      for await (const folder of foldersToDelete) {
```
Comment (ipapandinas) [inline src/commands/clear/index.ts:75]: `await` is useless in the for loop declaration, is already perform at the spinner command level

File: src/commands/clear/index.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        `Error deleting the folder ${path}.`, {cause: err}
+      );
+    }
+  } 
+
+  public async run(): Promise<any> {
+
+    const { flags, args } = await this.parse(Clear);
+
+    if(args.contractName === undefined && !flags.all) throw new ConfigError("You need to send any flag or argument.")
```
Comment (ipapandinas) [inline src/commands/clear/index.ts:40]: The config error message can be more explicit, something like: "Specify a contract name or use the --all flag to delete all artifacts."

File: src/commands/clear/index.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+    const workDirectory = process.cwd();
+    let foldersToDelete: Folder[] = []
+
+    if(flags.all) {
+
+      foldersToDelete.push({
```
Comment (ipapandinas) [inline src/commands/clear/index.ts:47]: No need to push, a direct `foldersToDelete` assignment works well:

```typescript
foldersToDelete = [
  { name: "Artifacts", path: path.join(workDirectory, './artifacts') },
  { name: "Target", path: path.join(workDirectory, './target') }
];
```

File: src/commands/clear/index.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+      })
+
+      for await (const folder of foldersToDelete) {
+        let resultSpinner = await this.spinner.runCommand( async () => this.deleteFolder(folder.path),
+         `Deleting the ${folder.name} folder`
+      )
+      } 
+    } else if (args.contractName) {
+
+      foldersToDelete.push({
```
Comment (ipapandinas) [inline src/commands/clear/index.ts:62]: Same here, you can make a direct `foldersToDelete` assignment too.

**Review by ipapandinas — CHANGES_REQUESTED** [review]
sorry for the late response @Juminstock, your implementation of the `clear` command looks pretty solid, thank you for contributing!

I added some comments for small code structure adjustments; once implemented I will approve this PR

**Comment by Juminstock** [comment]
Hi @ipapandinas, thank you for your comments in the code. I've resolved the bugs and improve the code based on your tips!

**Review by ipapandinas — APPROVED** [review]
LGTM, thank you for your contribution @Juminstock 🙌
It will be include in the upcoming new release completing the [ink-devhub-1](https://github.com/inkdevhub/swanky-cli/tree/ink-devhub-1) proposal.

**Pull request merged**